File: src/epub_generator.py
from ebooklib import epub
from typing import Optional
import os
from datetime import datetime
import re


class EPUBGenerator:
    """Generates EPUB files from text chunks."""

    def __init__(self, output_dir: str = "epubs"):
        self.output_dir = output_dir
        os.makedirs(output_dir, exist_ok=True)

    def create_epub(
        self,
        text: str,
        title: str,
        author: str = "Unknown Author",
        chunk_number: Optional[int] = None,
        total_chunks: Optional[int] = None,
    ) -> str:
        """
        Create an EPUB file from text.

        Args:
            text: Story text content
            title: Story title
            author: Author name
            chunk_number: Chunk number (if part of series)
            total_chunks: Total number of chunks

        Returns:
            Path to generated EPUB file
        """
        book = epub.EpubBook()

        # Add metadata
        book.set_identifier(f"{title}_{chunk_number or 1}_{datetime.utcnow().timestamp()}")

        if chunk_number and total_chunks:
            full_title = f"{title} - Part {chunk_number}/{total_chunks}"
        elif chunk_number:
            full_title = f"{title} - Part {chunk_number}"
        else:
            full_title = title

        book.set_title(full_title)
        book.set_language("en")
        book.add_author(author)

        # Create chapter
        chapter = epub.EpubHtml(
            title=full_title,
            file_name="chapter.xhtml",
            lang="en"
        )

        # Convert text to HTML (handles both plain text and HTML input)
        html_content = self._prepare_html_content(text)
        chapter.content = f"<h1>{full_title}</h1>{html_content}"

        # Add chapter to book
        book.add_item(chapter)

        # Define table of contents
        book.toc = (chapter,)

        # Add navigation files
        book.add_item(epub.EpubNcx())
        book.add_item(epub.EpubNav())

        # Define spine
        book.spine = ["nav", chapter]

        # Generate filename
        safe_title = self._sanitize_filename(title)
        if chunk_number:
            filename = f"{safe_title}_part{chunk_number}.epub"
        else:
            filename = f"{safe_title}.epub"

        filepath = os.path.join(self.output_dir, filename)

        # Write EPUB file
        epub.write_epub(filepath, book)

        return filepath

    def _prepare_html_content(self, text: str) -> str:
        """Prepare content for EPUB - handles both HTML and plain text."""
        # Check if content is already HTML
        if self._is_html(text):
            # Content is already HTML, use it directly
            return text
        else:
            # Plain text, convert to HTML paragraphs
            return self._text_to_html(text)

    def _is_html(self, text: str) -> bool:
        """Check if text contains HTML tags."""
        # Simple heuristic: if it has HTML tags, treat as HTML
        html_pattern = r'<(?:p|div|article|section|h[1-6]|br|span|em|strong|a)\b[^>]*>'
        return bool(re.search(html_pattern, text, re.IGNORECASE))

    def _text_to_html(self, text: str) -> str:
        """Convert plain text to HTML paragraphs."""
        paragraphs = text.split("\n\n")
        html_parts = []

        for para in paragraphs:
            para = para.strip()
            if para:
                # Preserve single line breaks within paragraphs
                para = para.replace("\n", "<br/>")
                html_parts.append(f"<p>{para}</p>")

        return "\n".join(html_parts)

    def _sanitize_filename(self, filename: str) -> str:
        """Sanitize filename to remove invalid characters."""
        # Remove or replace invalid filename characters
        invalid_chars = '<>:"/\\|?*'
        for char in invalid_chars:
            filename = filename.replace(char, "_")

        # Limit length
        max_length = 200
        if len(filename) > max_length:
            filename = filename[:max_length]

        return filename.strip()

    def create_multipart_epubs(
        self,
        chunks: list[tuple[str, int]],
        title: str,
        author: str = "Unknown Author",
    ) -> list[str]:
        """
        Create multiple EPUB files from text chunks.

        Args:
            chunks: List of (text, word_count) tuples
            title: Story title
            author: Author name

        Returns:
            List of paths to generated EPUB files
        """
        total_chunks = len(chunks)
        epub_paths = []

        for i, (chunk_text, _) in enumerate(chunks, start=1):
            epub_path = self.create_epub(
                text=chunk_text,
                title=title,
                author=author,
                chunk_number=i,
                total_chunks=total_chunks,
            )
            epub_paths.append(epub_path)

        return epub_paths

File: nextjs-app/lib/epub-generator.ts
/**
 * EPUB Generator
 *
 * Generate EPUB files from text content using epub-gen-memory.
 * Matches functionality of Python EPUBGenerator class.
 */

import { EPub } from "epub-gen-memory";

export interface EPUBOptions {
  title: string;
  author?: string;
  content: string;
  chunkNumber?: number;
  totalChunks?: number;
}

/**
 * Sanitize filename to remove invalid characters
 */
function sanitizeFilename(filename: string): string {
  const invalidChars = /[<>:"/\\|?*]/g;
  let sanitized = filename.replace(invalidChars, "_");

  // Limit length
  const maxLength = 200;
  if (sanitized.length > maxLength) {
    sanitized = sanitized.substring(0, maxLength);
  }

  return sanitized.trim();
}

/**
 * Check if text contains HTML tags
 */
function isHTML(text: string): boolean {
  const htmlPattern =
    /<(?:p|div|article|section|h[1-6]|br|span|em|strong|a)\b[^>]*>/i;
  return htmlPattern.test(text);
}

/**
 * Convert plain text to HTML paragraphs
 */
function textToHTML(text: string): string {
  const paragraphs = text.split("\n\n");
  const htmlParts: string[] = [];

  for (const para of paragraphs) {
    const trimmed = para.trim();
    if (trimmed) {
      // Preserve single line breaks within paragraphs
      const withBreaks = trimmed.replace(/\n/g, "<br/>");
      htmlParts.push(`<p>${withBreaks}</p>`);
    }
  }

  return htmlParts.join("\n");
}

/**
 * Prepare content for EPUB - handles both HTML and plain text
 */
function prepareHTMLContent(text: string): string {
  if (isHTML(text)) {
    // Content is already HTML, use it directly
    return text;
  } else {
    // Plain text, convert to HTML paragraphs
    return textToHTML(text);
  }
}

/**
 * Generate an EPUB file from text content
 *
 * @returns Buffer containing the EPUB file data
 */
export async function generateEPUB(options: EPUBOptions): Promise<Buffer> {
  const {
    title,
    author = "Unknown Author",
    content,
    chunkNumber,
    totalChunks,
  } = options;

  // Build full title with chunk info
  let fullTitle = title;
  if (chunkNumber && totalChunks) {
    fullTitle = `${title} - Part ${chunkNumber}/${totalChunks}`;
  } else if (chunkNumber) {
    fullTitle = `${title} - Part ${chunkNumber}`;
  }

  // Prepare HTML content
  const htmlContent = prepareHTMLContent(content);

  // Configure EPUB options
  const epubOptions = {
    title: fullTitle,
    author: author,
    publisher: "Nighttime Story Prep",
    lang: "en",
  };

  // Configure content
  const epubContent = [
    {
      title: fullTitle,
      content: `<h1>${fullTitle}</h1>${htmlContent}`,
    },
  ];

  try {
    // Generate EPUB in memory
    const epub = new EPub(epubOptions, epubContent);
    const buffer = await epub.genEpub();

    console.log(
      `[EPUB] Generated EPUB for "${fullTitle}" (${buffer.length} bytes)`
    );

    return buffer;
  } catch (error) {
    console.error("[EPUB] Failed to generate EPUB:", error);
    throw new Error(
      `Failed to generate EPUB: ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/**
 * Generate filename for EPUB
 */
export function generateEPUBFilename(
  title: string,
  chunkNumber?: number
): string {
  const safeTitle = sanitizeFilename(title);

  if (chunkNumber) {
    return `${safeTitle}_part${chunkNumber}.epub`;
  }

  return `${safeTitle}.epub`;
}

/**
 * Generate multiple EPUBs from chunk data
 */
export async function generateMultipleEPUBs(
  chunks: Array<{ text: string; wordCount: number }>,
  title: string,
  author?: string
): Promise<Array<{ buffer: Buffer; filename: string; chunkNumber: number }>> {
  const totalChunks = chunks.length;
  const epubs: Array<{
    buffer: Buffer;
    filename: string;
    chunkNumber: number;
  }> = [];

  for (let i = 0; i < chunks.length; i++) {
    const chunkNumber = i + 1;
    const chunk = chunks[i];

    const buffer = await generateEPUB({
      title,
      author,
      content: chunk.text,
      chunkNumber,
      totalChunks,
    });

    const filename = generateEPUBFilename(title, chunkNumber);

    epubs.push({
      buffer,
      filename,
      chunkNumber,
    });
  }

  return epubs;
}
